src/algebra/quandle.lean

```diff
@@ -0,0 +1,561 @@
+
+/--
+The map `x ↦ x ◃ x` is a bijection.  (This has applications for the
+regular isotopy version of the Reidemeister I move for knot diagrams.)
+-/
+def self_apply_equiv (R : Type*) [rack R] : R ≃ R :=
+{ to_fun := λ x, x ◃ x,
+  inv_fun := λ x, x ◃⁻¹ x,
+  left_inv := λ x, by simp,
+  right_inv := λ x, by simp }
+
+/--
+An involutory rack is one for which `rack.op R x` is an involution for every x.
+-/
+def is_involutory (R : Type*) [rack R] : Prop := ∀ x : R, function.involutive (op' x)
+
+lemma involutory_op_inv_eq_op {R : Type*} [rack R] (h : is_involutory R) (x y : R) :
+  x ◃⁻¹ y = x ◃ y :=
+begin
+  rw [←left_cancel x, right_inv],
+  exact((h x).left_inverse y).symm,
+end
+
+/--
+An abelian rack is one for which the mediality axiom holds.
+-/
+def is_abelian (R : Type*) [rack R] : Prop :=
+∀ (x y z w : R), (x ◃ y) ◃ (z ◃ w) = (x ◃ z) ◃ (y ◃ w)
+
+/--
+Associative racks are uninteresting.
+-/
+lemma assoc_iff_id {R : Type*} [rack R] {x y z : R} :
+  x ◃ y ◃ z = (x ◃ y) ◃ z ↔ x ◃ z = z :=
+by { rw self_distrib, rw left_cancel }
+
+end rack
+
+/--
+The type of homomorphisms between racks.
+This is also the notion of a quandle homomorphism.
+-/
+@[ext]
+structure rack_hom (R₁ : Type*) (R₂ : Type*) [rack R₁] [rack R₂] :=
+(to_fun : R₁ → R₂)
+(map_op' : ∀ {x y : R₁}, to_fun (x ◃ y) = to_fun x ◃ to_fun y)
+
+namespace rack_hom
+
+attribute [simp] rack_hom.map_op'
+
+instance (R₁ : Type*) (R₂ : Type*) [rack R₁] [rack R₂] : has_coe_to_fun (rack_hom R₁ R₂) :=
+⟨_, rack_hom.to_fun⟩
+
+@[simp]
+lemma map_op {R₁ : Type*} {R₂ : Type*} [rack R₁] [rack R₂] (f : rack_hom R₁ R₂)
+  (x y : R₁) : f (x ◃ y) = f x ◃ f y :=
+rack_hom.map_op' f
+
+/-- The identity homomorphism -/
+def id (R : Type*) [rack R] : rack_hom R R :=
+{ to_fun := id,
+  map_op' := by simp }
+
+instance inhabited (R : Type*) [rack R] : inhabited (rack_hom R R) :=
+⟨id R⟩
+
+variables {R₁ : Type*} {R₂ : Type*} {R₃ : Type*} [rack R₁] [rack R₂] [rack R₃]
+
+/-- The composition of rack/quandle homomorphisms -/
+def comp (g : rack_hom R₂ R₃) (f : rack_hom R₁ R₂) : rack_hom R₁ R₃ :=
+{ to_fun := g.to_fun ∘ f.to_fun,
+  map_op' := by simp }
+
+@[simp]
+lemma comp_apply (g : rack_hom R₂ R₃) (f : rack_hom R₁ R₂) (x : R₁) :
+  (g.comp f) x = g (f x) := rfl
+
+end rack_hom
+
+
+/--
+A quandle is a rack such that each automorphism fixes its corresponding element.
+-/
+class quandle (α : Type*) extends rack α :=
+(fix' : ∀ {x : α}, op' x x = x)
+
+namespace quandle
+open rack
+variables {Q : Type*} [quandle Q]
+
+@[simp]
+lemma fix {x : Q} : x ◃ x = x :=
+quandle.fix'
+
+lemma fix_inv {x : Q} : x ◃⁻¹ x = x :=
+by { rw ←left_cancel x, simp }
+
+instance opp.quandle : quandle (opp Q) :=
+{ fix' := λ x, fix_inv }
+
+/--
+The conjugation quandle of a group.  Each element of the group acts by
+the corresponding inner automorphism.
+-/
+@[nolint has_inhabited_instance]
+def conj (G : Type*) := G
+
+@[simp] lemma conj_conj_inv {G : Type*} [group G] (g h : G) : g * (g⁻¹ * h * g) * g⁻¹ = h :=
```

Comment:
I wasn't sure where to place these lemmas (or what to call them), so they are here for now.  They're used for `conj.quandle`.

Reply:
Does the `group` tactic suffice, instead of defining `simp` lemmas and calling `simp`?

Reply:
Thanks, didn't know about the group tactic.